src/cache.ts

```diff
@@ -0,0 +1,215 @@
+import { R2HTTPMetadata, R2ListOptions } from '@cloudflare/workers-types/2023-07-01';
+import { Context } from './context';
+import { b64Tou8, u8ToB64 } from './utils/base64';
+import { PRIVATE_TOKEN_ISSUER_DIRECTORY } from '@cloudflare/privacypass-ts';
+
+export const FAKE_DOMAIN_CACHE = 'cache.local';
+
+export const getDirectoryCache = async (): Promise<Cache> => {
+	return caches.open('response/issuer-directory');
+};
+
+export const DIRECTORY_CACHE_REQUEST = new Request(
+	`https://${FAKE_DOMAIN_CACHE}${PRIVATE_TOKEN_ISSUER_DIRECTORY}`
+);
+
+export const clearDirectoryCache = async (): Promise<boolean> => {
+	const cache = await getDirectoryCache();
+	return cache.delete(DIRECTORY_CACHE_REQUEST);
+};
+
+export type CacheElement<T> = { value: T; expiration: Date };
+
+interface ReadableCache {
+	read<T>(key: string, setValFn: (key: string) => Promise<CacheElement<T>>): Promise<T>;
+}
+
+const serialize = <T>(value: T): string => {
+	return JSON.stringify(value, (_key, value) => {
+		if (value instanceof Uint8Array) {
+			return `u8-${u8ToB64(value)}`;
+		}
+		if (value instanceof Date) {
+			return `date-${value.toJSON()}`;
+		}
+		return value;
+	});
+};
+
+const deserialize = <T>(value: string): T => {
+	return JSON.parse(value, (_key, value) => {
+		if (value.startsWith && value.startsWith('u8-')) {
+			return b64Tou8(value.slice('u8-'.length));
+		}
+		if (value.startsWith && value.startsWith('date-')) {
+			return new Date(value.slice('date-'.length));
+		}
+		return value;
```

Comment:
This allows for a type confusion attack where a literal value starting with `u8-` or `date-` can be forced into a non-string data type.

Can you check the `key` instead to infer the data type instead of the value?